src/python/dxpy/sugar/context.py

```diff
@@ -0,0 +1,191 @@
+        >>> with set_env(PYTHONPATH=''):
+        ...    print(os.environ["PYTHONPATH"])
+        <BLANKLINE>
+
+        Usage 3: Manipulate multiple variables
+        >>> myenv = {"PLUGINS_DIR": u'test/plugins', "PYTHONPATH": u'some/python/path'}
+        >>> with set_env(**myenv):
+        ...   print(os.environ["PLUGINS_DIR"])
+        ...   print(os.environ["PYTHONPATH"])
+        test/plugins
+        some/python/path
+    """
+    old_environ = dict(os.environ)
+    os.environ.update(environ)
+    try:
+        yield
+    finally:
+        os.environ.clear()
+        os.environ.update(old_environ)
+
+
+class cd:
+    """
+    Context manager for changing the current working directory
+
+    Args:
+        target_path (string): Optional, specify path to cd to
+        cleanup (boolean): Optional, specify if directory should be deleted after exiting context. Default is true if
+        the directory is newly created. Existing directories are never deleted.
+
+    Note:
+        If no args specified, cd() will create an arbitary temp dir and cd to it
+
+    Yields:
+        Upon entry, context will be set to the specified directory.
+        Upon exit, directory newly created with cleanup=True or directory created when no
+        args are specified is deleted.
+
+    Source: http://stackoverflow.com/questions/431684/how-do-i-cd-in-python
+
+    Examples:
+       with cd():
+           do_the_thing
+           # this will create a temp directory with a randomly
+           # generated name, doe the thing, then delete the temp dir
+
+       with cd(my_file_dir):
+           do_the_thing
+           # this will do the thing in my_file_dir and not delete the directory
+
+       with cd(target_path=my_temp_dir, cleanup=True):
+           do_the_thing
+           # this will create a temp dir with path my_temp_dir, do the thing,
+           # then delete the temp dir
+    """
+
+    def __init__(self, target_path=None, cleanup=True):
+        if target_path is not None:
+            if os.path.exists(target_path):
+                self.newPath = target_path
+                self.removeFolder = False
+            else:
+                self.newPath = tempfile.mkdtemp(dir=target_path)
+                self.removeFolder = cleanup
+        else:
+            self.newPath = tempfile.mkdtemp()
+            self.removeFolder = cleanup
+
+    def __enter__(self):
+        self.savedPath = os.getcwd()
+        os.chdir(self.newPath)
+
+    def __exit__(self, etype, value, traceback):
+        os.chdir(self.savedPath)
+        if self.removeFolder:
+            subprocess.check_call(['rm', '-rf', self.newPath], shell=False)
+
+@contextlib.contextmanager
+def fifo(name=None):
+    """
+    Create a FIFO, yield it, and delete it before exiting.
+    Args:
+        name: The name of the FIFO, or None to use a temp name.
+    Yields:
+        The name of the FIFO
+    """
+    if name is None:
+        temp = tempfile.NamedTemporaryFile(delete=False)
```

Comment:
What is the difference between this and tempfile.mkstemp()[1]?

Reply:
No technical difference, NamedTemporaryFile calls mkstemp() on the backend.